cpp/oneapi/dal/backend/primitives/select.cpp

```diff
@@ -0,0 +1,249 @@
+                // Odd-Even Sorting
+                {
+                    // Sorting itself
+                    for (idx_t step = 0; step < k_width; ++step) {
+                        for (idx_t i = 2 * lxid + step & 1; i + 1 < k_width; i += xrange) {
+                            const bool swap = dwspace[lrid + i] > dwspace[lrid + i + 1];
+                            conditional_swap<Float>(dwspace[lrid + i], dwspace[lrid + i + 1], swap);
+                            conditional_swap<idx_t>(iwspace[lrid + i], iwspace[lrid + i + 1], swap);
+                        }
+                        idx.barrier(cl::sycl::access::fence_space::local_space);
+                    }
+                }
+            }
+        }
+
+        //Exit from algorithm
+        {
+            if (gyid < n_vectors) {
+                const idx_t* const irow = &(output_indices[gyid * k]);
+                const Float* const drow = &(ouput_distances[gyid * k]);
+                for (idx_t shift = lxid; shift < k; shift += xrange) {
+                    irow[shift] = iwspace[lrid + shift];
+                    drow[shift] = dwspace[lrid + shift];
+                }
+            }
+            // TODO: Optimize with async_work_group_copy
+        }
+    }
+
+private:
+    const Float* const input_cross;
+    const Float* const input_norms;
+    idx_t* const output_indices;
+    Float* const output_distances;
+    const idx_t k, n_vectors, k_width, vector_size;
+    dist_acc_t dwspace;
+    idxs_acc_t iwspace;
+};
+
+} // namespace impl
+
+template <typename Float>
+select_small_k_l2<Float>::select_small_k_l2(cl::sycl::queue& queue)
+        : q(queue),
+          max_work_group_size(
+              queue.get_device().template get_info<cl::sycl::info::device::max_work_group_size>()),
+          max_local_size(
+              queue.get_device().template get_info<cl::sycl::info::device::local_mem_size>() /
+              elem_size),
+          preferred_width(queue.get_device().template get_info<preferred_size_flag>()) {}
+
+template <typename Float>
+std::pair<std::int64_t, std::int64_t> select_small_k_l2<Float>::preferred_local_size(const std::int64_t k) {
+    if (k > max_local_size)
+        throw std::exception();
+    const idx_t min_k_width    = preferred_size * (k / preferred_size + 1);
+    const idx_t yrange_mem_lim = max_local_size / min_k_width;
+    const idx_t yrange_max_lim = std::min(yrange_mem_lim, max_work_group_size);
+    const idx_t yrange_min_lim =
+        std::min(yrange_mem_lim, std::min(prefered_size, max_work_group_size));
+    idx_t yrange = yrange_min_lim, k_width = min_k_width;
+    for (; yrange < yrange_max_lim; yrange++) {
+        // Optimal k_width
+        k_width = max_local_size / yrange;
+        if (k_width % preffered_size == 0)
+            break;
+    }
+    if (yrange * k_width > max_local_size)
+        throw std::exception();
+    if (yrange > max_work_group_size)
+        throw std::exception();
```

Comment:
Don't use exceptions here, especially std. It's internal functionality ASSERT is enough